controllers/clustertopology_controller_compute.go

```diff
@@ -308,6 +332,101 @@ func computeCluster(current *clusterTopologyState, infrastructureCluster, contro
+	className := machineDeploymentTopology.Class
+	machineDeploymentClass := getMachineDeploymentClass(class, className)
+
+	machineDeploymentTopologyClass, ok := class.machineDeployments[className]
+	if !ok {
+		return state, errors.Errorf("class name %s not found", className)
+	}
+
+	state.bootstrapTemplate = templateToTemplate(templateToInput{
+		template:              machineDeploymentTopologyClass.bootstrapTemplate,
+		templateClonedFromRef: objToRef(machineDeploymentTopologyClass.bootstrapTemplate),
+		cluster:               current.cluster,
+		namePrefix:            fmt.Sprintf("%s-%s-bootstrap-", current.cluster.Name, machineDeploymentTopology.Name),
+		currentObjectRef:      machineDeploymentClass.Template.Bootstrap.Ref,
+		labels:                mergeMap(machineDeploymentTopology.Metadata.Labels, machineDeploymentClass.Template.Metadata.Labels),
+		annotations:           mergeMap(machineDeploymentTopology.Metadata.Annotations, machineDeploymentClass.Template.Metadata.Annotations),
+	})
+
+	state.infrastructureMachineTemplate = templateToTemplate(templateToInput{
+		template:              machineDeploymentTopologyClass.infrastructureMachineTemplate,
+		templateClonedFromRef: objToRef(machineDeploymentTopologyClass.infrastructureMachineTemplate),
+		cluster:               current.cluster,
+		namePrefix:            fmt.Sprintf("%s-%s-infra-", current.cluster.Name, machineDeploymentTopology.Name),
+		currentObjectRef:      machineDeploymentClass.Template.Infrastructure.Ref,
+		labels:                mergeMap(machineDeploymentTopology.Metadata.Labels, machineDeploymentClass.Template.Metadata.Labels),
+		annotations:           mergeMap(machineDeploymentTopology.Metadata.Annotations, machineDeploymentClass.Template.Metadata.Annotations),
+	})
+
+	machineDeployment := computeMachineDeployment(current, currentMachineDeployment, machineDeploymentTopology, machineDeploymentClass)
+	machineDeployment.Spec.Template.Spec.Bootstrap.ConfigRef = objToRef(state.bootstrapTemplate)
+	machineDeployment.Spec.Template.Spec.InfrastructureRef = *objToRef(state.infrastructureMachineTemplate)
+	state.object = machineDeployment
+
+	return state, nil
+}
+
+// computeMachineDeployment computes a MachineDeployment object for a topology defined in the Cluster object
+// The generated machineDeployment object is calculated using the values from the machineDeploymentTopology and
+// the machineDeployment class.
+func computeMachineDeployment(cluster *clusterTopologyState, current *clusterv1.MachineDeployment, topology clusterv1.MachineDeploymentTopology, class *clusterv1.MachineDeploymentClass) *clusterv1.MachineDeployment {
+	gv := clusterv1.GroupVersion
+	machineDeployment := &clusterv1.MachineDeployment{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       gv.WithKind("MachineDeployment").Kind,
+			APIVersion: gv.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-%s-", cluster.cluster.Name, topology.Name)),
```

Comment:
Do we want to / have to hash the name. Proposal:
```bash
1. Create a name `<cluster-name>-<worker-set-name>` (if too long, hash it)
```

(this would also affect the BootstrapTemplate and InfrastructureMachineTemplate names above)

Mainly asking now already because this also affects my reconcile MachineDeployment PR which builds on top.

Reply:
It is not required to hash the name because `GenerateName` already ensure the name does not exceeds max length.

Also reconcille MachineDeployment should not rely on metadata.Name but it should rely on the label with the topology name